### PR TITLE
feat: implement EIP-7685

### PR DESCRIPTION
**Pull request by onbjerg**
Adds types for EIP-7685, adds a new `Requests` table, and adjusts the code where appropriate.

This still lacks:

- Blockchain tree integration
- Writing the `Requests` in the bodies stage
- Validating the requests root in the body downloader
- Full engine API integration

Will do these in separate PRs to unblock EIPs 6110 and 7002

### CODE REVIEW

File: crates/consensus/auto-seal/src/lib.rs [hunk 1]
```diff
@@ -325,6 +326,8 @@ impl StorageInner {
                 Some(calculate_excess_blob_gas(parent_excess_blob_gas, parent_blob_gas_used))
         }
 
+        // todo: do we support requests root in auto-seal
```
Comment (gakonst) [inline crates/consensus/auto-seal/src/lib.rs:329]: Just leave it empty

File: crates/primitives/src/proofs.rs [hunk 2]
```diff
@@ -72,6 +72,11 @@ pub fn calculate_receipt_root(receipts: &[ReceiptWithBloom]) -> B256 {
     ordered_trie_root_with_encoder(receipts, |r, buf| r.encode_inner(buf, false))
 }
 
+/// Calculate EIP-7685 requests root.
+pub fn calculate_requests_root(requests: &Requests) -> B256 {
```
Comment (mattsse) [inline crates/primitives/src/proofs.rs:76]: let's do &[Request] here

File: crates/primitives/src/request.rs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+//! EIP-7685 requests.
+
+use alloy_consensus::Request;
+use alloy_rlp::{RlpDecodableWrapper, RlpEncodableWrapper};
+use reth_codecs::{main_codec, Compact};
+
+/// A list of EIP-7685 requests.
+#[main_codec]
+#[derive(Debug, Clone, PartialEq, Eq, Default, Hash, RlpEncodableWrapper, RlpDecodableWrapper)]
+pub struct Requests(pub Vec<Request>);
```
Comment (mattsse) [inline crates/primitives/src/request.rs:10]: we're missing a few helper impls but this is ok

**Review by mattsse — APPROVED** [review]
smol nit

File: crates/rpc/rpc/src/eth/api/pending_block.rs [hunk 4]
```diff
@@ -244,6 +245,15 @@ impl PendingBlockEnv {
...
 
+        // note(onbjerg): the rpc spec has not been changed to include requests, so for now we just
+        // set these to empty
+        let (requests, requests_root) =
+            if chain_spec.is_prague_active_at_timestamp(block_env.timestamp.to::<u64>()) {
+                (Some(Requests::default()), Some(EMPTY_ROOT_HASH))
+            } else {
+                (None, None)
+            };
+
```
Comment (gakonst) [inline crates/rpc/rpc/src/eth/api/pending_block.rs:256]: https://github.com/ethereum/execution-apis/pull/545

File: crates/revm/src/batch.rs [hunk 5]
```diff
@@ -84,7 +84,7 @@ impl BlockBatchRecord {
     }
 
     /// Returns all recorded requests.
-    pub fn take_requests(&mut self) -> Requests {
+    pub fn take_requests(&mut self) -> Vec<Requests> {
```
Comment (gakonst) [inline crates/revm/src/batch.rs:87]: I think we start having a lot of these `Vec<Vec<_>>` outer / inner vecs so maybe we should do a generic `Batch<T>` to encapsulate that

File: crates/primitives/src/block.rs [hunk 6]
```diff
@@ -13,6 +13,16 @@ pub use alloy_eips::eip1898::{
     BlockHashOrNumber, BlockId, BlockNumHash, BlockNumberOrTag, ForkBlock, RpcBlockHash,
 };
 
+// HACK(onbjerg): we need this to always set `requests` to `None` since we might otherwise generate
+// a block with `None` withdrawals and `Some` requests, in which case we end up trying to decode the
+// requests as withdrawals
```
Comment (gakonst) [inline crates/primitives/src/block.rs:18]: this seems fine, not a hack?

File: crates/rpc/rpc/src/eth/api/pending_block.rs (hunk 4 above)
Comment (onbjerg) [reply crates/rpc/rpc/src/eth/api/pending_block.rs:256]: this is the engine api, not the json rpc

File: crates/storage/provider/src/providers/database/provider.rs [hunk 7]
```diff
@@ -727,6 +736,7 @@ impl<TX: DbTxMut + DbTx> DatabaseProvider<TX> {
         let block_ommers = self.get_or_take::<tables::BlockOmmers, TAKE>(range.clone())?;
         let block_withdrawals =
             self.get_or_take::<tables::BlockWithdrawals, TAKE>(range.clone())?;
+        let block_requests = self.get_or_take::<tables::BlockRequests, TAKE>(range.clone())?;
```
[the remark this answers is not in the document]
Comment (onbjerg) [reply crates/storage/provider/src/providers/database/provider.rs:739]: agree

**Review by mattsse — APPROVED** [review]

**Pull request merged**